src/crypto_helper/algorithm.rs

```diff
@@ -361,6 +363,125 @@ pub struct ZlibInput {
     #[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")]
     pub data: Vec<u8>,
 }
+#[derive(Eq, Clone, Copy, PartialEq, Debug, Serialize, Deserialize, Default)]
+pub enum Argon2Variant {
+    Argon2i,
+    Argon2d,
+    #[default]
+    Argon2id,
+}
+#[derive(Eq, Clone, Copy, PartialEq, Debug, Serialize, Deserialize, Default)]
+pub enum Argon2Version {
+    Version10,
+    #[default]
+    Version13,
+}
+#[derive(Eq, Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub struct Argon2HashAction {
+    pub memory: u32,
+    pub iters: u32,
+    pub paralelism: u32,
+    pub output_len: u32,
+    pub salt: Vec<u8>,
+    pub data: Vec<u8>,
+    pub secret: Vec<u8>,
+    pub variant: Argon2Variant,
+    pub version: Argon2Version,
+}
+
+#[derive(Eq, Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub enum Argon2Action {
+    Hash(Argon2HashAction),
+    Verify(Vec<u8>),
+}
+#[derive(Eq, Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub struct Argon2Input {
+    pub action: Argon2Action,
+    #[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")]
+    pub data: Vec<u8>,
+}
+
+impl From<Argon2Variant> for argon2::Algorithm {
+    fn from(value: Argon2Variant) -> Self {
+        match value {
+            Argon2Variant::Argon2d => argon2::Algorithm::Argon2d,
+            Argon2Variant::Argon2i => argon2::Algorithm::Argon2i,
+            Argon2Variant::Argon2id => argon2::Algorithm::Argon2id,
+        }
+    }
+}
+
+impl From<Argon2Version> for argon2::Version {
+    fn from(value: Argon2Version) -> Self {
+        match value {
+            Argon2Version::Version10 => argon2::Version::V0x10,
+            Argon2Version::Version13 => argon2::Version::V0x13,
+        }
+    }
+}
+
+impl From<&Argon2HashAction> for argon2::Params {
+    fn from(value: &Argon2HashAction) -> Self {
+        argon2::ParamsBuilder::new()
+            .m_cost(value.memory)
+            .p_cost(value.paralelism)
+            .t_cost(value.iters)
+            .build()
+            .unwrap()
+    }
+}
+
+impl<'a> TryFrom<&'a Argon2HashAction> for argon2::password_hash::Salt<'a> {
+    type Error = String;
+    fn try_from(hash_action: &'a Argon2HashAction) -> Result<Self, Self::Error> {
+        argon2::password_hash::Salt::from_b64(
+            std::str::from_utf8(&hash_action.salt).map_err(|_| "Non UTF-8 salt".to_owned())?,
```

Comment:
salt should not be UTF-8. salt is just random bytes.
you should encode them as base64 and then pass them into `Salt::from_b64`

Reply:
But we need to somehow get a string. Personally, I don't know the way to get a non-utf8 string in rust (besides [`std::from_utf8_lossy`](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.from_utf8_lossy), or this is a sound solution? ).

Reply:
1. you store salt as `Vec<u8>`
2. when you need to provide it to the `argon2`, then do the following: `Salt::from_b64(base64::encode(&hash_action.salt))`

Reply:
Fixed 🫡